addons/l10n_ro_efactura/models/res_company.py

```diff
@@ -0,0 +1,99 @@
+import base64
+import requests
+
+from datetime import datetime
+from dateutil.relativedelta import relativedelta
+from werkzeug.urls import url_join
+
+from odoo import fields, models, api, _
+from odoo.exceptions import UserError
+from odoo.tools import json
+
+
+class ResCompany(models.Model):
+    _inherit = 'res.company'
+
+    l10n_ro_edi_client_id = fields.Char(string='Client ID')
+    l10n_ro_edi_client_secret = fields.Char(string='Client Secret')
+    l10n_ro_edi_access_token = fields.Char(string='Access Token')
+    l10n_ro_edi_refresh_token = fields.Char(string='Refresh Token')
+    l10n_ro_edi_access_expiry_date = fields.Date(string='Access Token Expiry Date')
+    l10n_ro_edi_refresh_expiry_date = fields.Date(string='Refresh Token Expiry Date')
+    l10n_ro_edi_callback_url = fields.Char(compute='_compute_l10n_ro_edi_callback_url')
+    l10n_ro_edi_test_env = fields.Boolean(string='Use Test Environment', default=True)
+    l10n_ro_edi_oauth_error = fields.Char()  # Error field to be shown in case of error from the authentication process
+
+    @api.depends('country_code')
+    def _compute_l10n_ro_edi_callback_url(self):
+        """ Callback URLs are used for generating client_id and client_secret from l10n_ro_edi's setting. """
+        for company in self:
+            if company.country_code == 'RO':
+                company.l10n_ro_edi_callback_url = url_join(company.get_base_url(), 'l10n_ro_edi/callback/%s' % company.id)
```

Comment:
That's my first time hearing this case (odoo with separate website).

@odoo/rd-website would it be possible?

The reason I'm doing this is that:

The government's API requires us to give a callback URL, which they will use to check if the request URL match what was given to them.
So we had to create a unique URL (that we can access to) which we'll use for the authentication+token generation

Reply:
Hello :wave: 

I am definitely no python things expert and not even a website expert when it comes to multi-company + multi-website things [but I'll try to be](https://www.youtube.com/watch?v=BKorP55Aqvg) (@rdeodoo was the one investigating those kinds of questions but he is gone now :sob:).

> if the company has a separate website but is interacted with from the common backend I don't know what the results will be when redirecting to the website backend
> [...]
> @odoo/rd-website would it be possible?

A website is forcefully linked to a company. A company may be linked to no website, one website or multiple websites. But in the case of multiple websites, we have a `website_id` field which links to the "main" website of the company. This is that "main" website base URL that will be used when doing `company.get_base_url()`.

I'll need to know more to answer more than that :shrug: Indeed, if you are logged in "secondary website", call some kind of API from the backend with a callback using the "main website" base URL so that the API can redirect you to it, you'll end up on the "main website", where you are potentially not logged in I guess.

Some other code seems to do this kind of things though :shrug: I guess given the way things work, it needs to be a choice between:
- "That's fine if you use some main company website URL at some point" (in this case, to be redirected to)
- Doing the action can only be done from "the main website backend" (kinda like for website edition: if you are logged in the backend using "website 1" URL, if you try to edit "website 2", you are redirected first (and potentially need to login again). Probably checking the current URL via `request` and comparing it to `company.get_base_url()`? :shrug: 

I am mostly guessing here. Feel free to rescue me @rdeodoo @JKE-be :shrug:

Reply:
`if request and request.httprequest.url_root` payment is doing that in their own override of `get_base_url()`, couldn't this `url_root` also be used in this case to reach back the URL from where the request originated?
I did not read the PR and I don't know who/how is the call to third party performed, but I guess that if you want to reach back to a specific website it means that it's an end user flow, in which case there should be an `url_root` and it would make sense to come back to this URL?